Feature/add glossary page

BonfaceKilz opened the pull request:
#### Description
<!--Brief description of the PR. What does this PR do? -->
This PR:
- Isolates integration tests from unittests
- Adds a glossary that's rendered from Markdown
- Adds basic CSS for the Markdown file
- Replaces references to the glossary with the right link

#### How should this be tested?

<!-- What should you do to test this PR? Is there any manual quality
assurance checks that should be done. What are the expectations -->
Go to (temporarily set up for this PR) http://gn2-test1.genenetwork.org/glossary/

#### Any background context you want to provide?

<!-- Anything the reviewer should be aware of ahead of testing -->
- The idea is to demo a working version of using md files to display and edit content with the end goal of making contributing easier.
- You can use valid html inside markdown files

#### What are the relevant pivotal tracker stories?

<!-- Does this PR track anything anywhere? -->
resolves https://github.com/genenetwork/genenetwork2/issues/372

#### Screenshots (if appropriate)

<details>
<summary>Click to see full image</summary>
<img src="https://user-images.githubusercontent.com/11820306/98805954-78919d00-2429-11eb-9e56-4290429e01b0.png" alt="screenshot"/>
</details>

#### Questions

<!-- Are there any questions for the reviewer -->
I could split up the md files, but would that really be useful?

BonfaceKilz committed:
Apply PEP-8 to file

* wqflask/wqflask/__init__.py: Remove unused import and
variables. Also add module docstring and pylint errors to ignore.
(before_request): Use correct spacing.

BonfaceKilz committed:
Add and register glossary blueprint

BonfaceKilz committed:
Add glossary template and use it

BonfaceKilz committed:
Add test for "/glossary" route

BonfaceKilz committed:
Separate unittests from integration tests

BonfaceKilz committed:
Remove encoding header for file

In python3 the default encoding is utf-8 so this is redundant.

BonfaceKilz committed:
Rename glossary.py to markdown_routes.py

Evary page with markdown content will have routes added in
markdown_routes.py

* markdown_routes.py: New file.
* wqflask/wqflask/glossary.py: Delete it.
* wqflask/wqflask/__init__.py: Update import.

BonfaceKilz committed:
Update glossary template with markdown content

BonfaceKilz committed:
Add simple md content in glossary.md

BonfaceKilz committed:
Use python3-genenetwork2 on commit bdce85d in container

* .github/workflows/main.yml: Update gn2 container.

BonfaceKilz committed:
Move logic for fetching md files to it's own function

* wqflask/wqflask/markdown_routes.py
(render_markdown): New function.
(glossary): use render_markdown function.

BonfaceKilz committed:
Add basic structure for "/glossary" routes test

BonfaceKilz committed:
Add basic tests for rendering_markdown

* wqflask/tests/unit/wqflask/test_markdown_routes.py: New tests.

BonfaceKilz committed:
Delete file

* wqflask/tests/integration/test_glossary.py: Delete it. Earlier
renamed to test_markdown_routes.

BonfaceKilz committed:
Fix false-positive tests

* wqflask/tests/unit/wqflask/test_markdown_routes.py:
(MockRequests404): Pass self in all properties.
(MockRequests200): Ditto.
(test_render_markdown): Rename to
test_render_markdown_when_fetching_locally.
(test_render_markdown_when_fetching_remotely): New test.

zsloan commented:
I haven't used Flask Blueprints before, but from reading about it some it seems like it can be a good idea to use them in situations where something is essentially separate from the rest of the web application (which is definitely the case for something like the glossary).

Are the advantages just things like lacking the overhead of having to actually load views.py and being able to have a separate @before_request/@after_request, etc? I'm wondering if it might make sense to use a Blueprint for things like user registration/authentication.

Alexanderlacuna commented:
I guess you are using  mistune==0.8.4  or less  and not mistune==2.0.0a4 .I i  inspected both
only the mistune=2.0.0a4 has a html attribute

Alexanderlacuna commented on wqflask/wqflask/markdown_routes.py:
```diff
@@ -0,0 +1,38 @@
+    with open(os.path.join(os.path.abspath(os.path.dirname(__file__)),
+                           f"static/markdown/{file_name}")) as md_file:
+        markdown = md_file.read()
+        return mistune.html(markdown)
```
Are you using mistune==0.8.4 or below try using mistune>=2.0.0a4 that might 
fix the bug you are having

BonfaceKilz committed:
Use python3-genenetwork2 docker image on commit 0bf4ee6

BonfaceKilz commented:
Just pushed python3-genenetwork3:0bf4ee6 which has the latest python-mistune-2-x-x [here](https://hub.docker.com/repository/docker/bonfacekilz/python3-genenetwork2)

BonfaceKilz commented:
zsloan <notifications@github.com> writes:

> I haven't used Flask Blueprints before, but from
> reading about it some it seems like it can be a
> good idea to use them in situations where
> something is essentially separate from the rest of
> the web application (which is definitely the case
> for something like the glossary).
>

Yeah that's the gist. But not entirely the whole
story. See below for more.

> Are the advantages just things like lacking the
> overhead of having to actually load views.py and
> being able to have a separate
> @before_request/@after_request, etc? I'm wondering
> if it might make sense to use a Blueprint for
> things like user registration/authentication.

Blueprints AFAIU provide ways to cleanly isolate
routes. One important use-case for us is that it
prevents unnecessary circular imports by
preventing the code author from passing around the
instantiated flask object(in our cases "app") like
what's happening in this file:
https://github.com/genenetwork/genenetwork2/blob/testing/wqflask/wqflask/user_session.py.

Also, since we register routes in one central
place(in our case, it'd be
https://github.com/genenetwork/genenetwork2/blob/testing/wqflask/wqflask/__init__.py),
it makes it easier to grow the flask code since
you can define routes without having to worry
about passing around some global object.

Generally as flask apps grow, IMHO blueprints are
the way to go; but more as a useful design tool.

HTH!

-- 
Bonface M. K. <https://www.bonfacemunyoki.com>
Chief Emacs Bazu / Rieng ya software sare
Mchochezi of: <https://upbookclub.com> / Twitter: @BonfaceKilz
GPG Key: D4F09EB110177E03C28E2FE1F5BBAE1E0392253F

BonfaceKilz committed:
Update glossary markdown file

BonfaceKilz committed:
Wrap rendered content in div container

BonfaceKilz committed:
Add an edit link button to glossary page

BonfaceKilz committed:
Update test_markdown_routes

* wqflask/tests/unit/wqflask/test_markdown_routes.py
(test_render_markdown_when_fetching_locally): Fix failing test.

BonfaceKilz committed:
Add link markers to elements which are referenced from elsewhere

BonfaceKilz committed:
Add missing image

BonfaceKilz committed:
Use new glossary link from glossary_blueprint in html templates

BonfaceKilz commented:
@robwwilliams have a look at: http://gn2-test1.genenetwork.org/glossary/. I may have to work on the styling and fix the links. I copy-pasted text and formatted it manually to md(programming it was way too much work imho). I'll brush things before I merge(but if I miss something, now you can edit it vis-a-vis github!). Also, the actual md file is: https://github.com/BonfaceKilz/genenetwork2/blob/feature/add-glossary-page/wqflask/wqflask/static/markdown/glossary.md#index. Would that be something you'd be comfortable working with?

zsloan commented:
The only minor thing that I think Rob will probably agree with is bolding
the glossary terms (though Rob will also be able to do stuff like that
himself through editing the md file on Github, so it's not that important).

On Wed, Nov 4, 2020 at 12:44 PM BonfaceKilz <notifications@github.com>
wrote:

> @robwwilliams <https://github.com/robwwilliams> have a look at:
> http://gn2-test1.genenetwork.org/glossary/. I may have to work on the
> styling and fix the links. I copy-pasted text and formatted it manually to
> md(programming it was way too much work imho). I'll brush things before I
> merge(but if I miss something, now you can edit it vis-a-vis github!).
> Also, the actual md file is:
> https://github.com/BonfaceKilz/genenetwork2/blob/feature/add-glossary-page/wqflask/wqflask/static/markdown/glossary.md#index.
> Would that be something you'd be comfortable working with?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork2/pull/477#issuecomment-721906903>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANQJGCMDNPPWKGCOMDTIBTSOGOJDANCNFSM4TJFLRLA>
> .
>

BonfaceKilz commented:
zsloan <notifications@github.com> writes:

> The only minor thing that I think Rob will probably agree with is bolding
> the glossary terms (though Rob will also be able to do stuff like that
> himself through editing the md file on Github, so it's not that important).
>

Ah yes! I'm thinking of just adding a markdown.css
file that handles that(in terms of styling) for
us. I'll do that sometime tomorrow though. WDYT?

-- 
Bonface M. K. <https://www.bonfacemunyoki.com>
Chief Emacs Bazu / Rieng ya software sare
Mchochezi of: <https://upbookclub.com> / Twitter: @BonfaceKilz
GPG Key: D4F09EB110177E03C28E2FE1F5BBAE1E0392253F

robwwilliams commented:
Bonface: this looks great.

At least one image did not load for me; perhaps lost at some point.

Hurray re edit in GitHub. That is big.

Thanks a ton.

On Wed, Nov 4, 2020 at 12:44 PM BonfaceKilz <notifications@github.com>
wrote:

> @robwwilliams <https://github.com/robwwilliams> have a look at:
> http://gn2-test1.genenetwork.org/glossary/. I may have to work on the
> styling and fix the links. I copy-pasted text and formatted it manually to
> md(programming it was way too much work imho). I'll brush things before I
> merge(but if I miss something, now you can edit it vis-a-vis github!).
> Also, the actual md file is:
> https://github.com/BonfaceKilz/genenetwork2/blob/feature/add-glossary-page/wqflask/wqflask/static/markdown/glossary.md#index.
> Would that be something you'd be comfortable working with?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork2/pull/477#issuecomment-721906903>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC55V7HGTJXVCNMTAN4EUVDSOGOJDANCNFSM4TJFLRLA>
> .
>


-- 
Rob

Robert W. Williams, Ph.D.
Chair: Department of Genetics, Genomics and Informatics
University of Tennessee Health Science Center
71 S Manassas St, Memphis TN 38103-3310,  TSRB Room 405
Tele/Cell: 901 604 4752
EMAIL: rwilliams@uthsc.edu  or  labwilliams@gmail.com

BonfaceKilz commented:
robwwilliams <notifications@github.com> writes:

> Bonface: this looks great.
>
> At least one image did not load for me; perhaps lost at some point.
>

Ouch! I'll fix that tomorrow. Probably I never
versioned it.

> Hurray re edit in GitHub. That is big.
>

Thanks!

> Thanks a ton.
>
Sure :)

-- 
Bonface M. K. <https://www.bonfacemunyoki.com>
Chief Emacs Bazu / Rieng ya software sare
Mchochezi of: <https://upbookclub.com> / Twitter: @BonfaceKilz
GPG Key: D4F09EB110177E03C28E2FE1F5BBAE1E0392253F

robwwilliams commented:
Agreed. I can update some of these terms and we can embedded some more
links. Soon it will be time for me to make a series of very short Guides to
GN.

On Wed, Nov 4, 2020 at 1:40 PM BonfaceKilz <notifications@github.com> wrote:

> zsloan <notifications@github.com> writes:
>
> > The only minor thing that I think Rob will probably agree with is bolding
> > the glossary terms (though Rob will also be able to do stuff like that
> > himself through editing the md file on Github, so it's not that
> important).
> >
>
> Ah yes! I'm thinking of just adding a markdown.css
> file that handles that(in terms of styling) for
> us. I'll do that sometime tomorrow though. WDYT?
>
> --
> Bonface M. K. <https://www.bonfacemunyoki.com>
> Chief Emacs Bazu / Rieng ya software sare
> Mchochezi of: <https://upbookclub.com> / Twitter: @BonfaceKilz
> GPG Key: D4F09EB110177E03C28E2FE1F5BBAE1E0392253F
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork2/pull/477#issuecomment-721935370>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC55V7B25ZVS25WFZ26UBJTSOGU4DANCNFSM4TJFLRLA>
> .
>


-- 
Rob

Robert W. Williams, Ph.D.
Chair: Department of Genetics, Genomics and Informatics
University of Tennessee Health Science Center
71 S Manassas St, Memphis TN 38103-3310,  TSRB Room 405
Tele/Cell: 901 604 4752
EMAIL: rwilliams@uthsc.edu  or  labwilliams@gmail.com

BonfaceKilz committed:
Add missing links and some markdown formatting(underline, bold)

* wqflask/wqflask/static/markdown/glossary.md: Update file.

BonfaceKilz committed:
Add basic markdown styling

* wqflask/wqflask/static/new/css/markdown.css: New file.
* wqflask/wqflask/templates/glossary.html: add markdown id to markdown
container.

BonfaceKilz committed:
Update image width

BonfaceKilz commented:
@robwwilliams this ready to be merged. I added missing links, correct highlighting(italics, bolds, underlines). Please see: http://gn2-test1.genenetwork.org/glossary/. I'll merge this later tonight.

PS: When pushing to production, you'd need to rebuild the gn2 package since I added python-mistune as an input [here](http://git.genenetwork.org/guix-bioinformatics/guix-bioinformatics/commit/4f121d7d2e397adbd227f67d38339f94dc26c707)

BonfaceKilz committed:
Merge branch 'testing' of github.com:genenetwork/genenetwork2 into feature/add-glossary-page

robwwilliams commented:
This looks beautiful Bonface. Thanks a ton. And now to edit in GitHub!

On Wed, Nov 11, 2020 at 5:35 AM BonfaceKilz <notifications@github.com>
wrote:

> @robwwilliams <https://github.com/robwwilliams> this ready to be merged.
> I added missing links, correct highlighting(italics, bolds, underlines).
> Please see: http://gn2-test1.genenetwork.org/glossary/. I'll merge this
> later tonight.
>
> PS: When pushing to production, you'd need to rebuild the gn2 package
> since I added python-mistune as an input here
> <http://git.genenetwork.org/guix-bioinformatics/guix-bioinformatics/commit/4f121d7d2e397adbd227f67d38339f94dc26c707>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork2/pull/477#issuecomment-725373425>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC55V7C5TTXEGH24LGMZ3NTSPJZJ3ANCNFSM4TJFLRLA>
> .
>


-- 
Rob

Robert W. Williams, Ph.D.
Chair: Department of Genetics, Genomics and Informatics
University of Tennessee Health Science Center
71 S Manassas St, Memphis TN 38103-3310,  TSRB Room 405
Tele/Cell: 901 604 4752
EMAIL: rwilliams@uthsc.edu  or  labwilliams@gmail.com
